jetty-server/src/main/java/org/eclipse/jetty/server/HttpOutput.java

```diff
@@ -673,11 +684,109 @@ public void write(int b) throws IOException
 
     @Override
     public void print(String s) throws IOException
+    {
+        print(s,false);
+    }
+
+    @Override
+    public void println(String s) throws IOException
+    {
+        print(s,true);
+    }
+
+    private void print(String s, boolean eoln) throws IOException
     {
         if (isClosed())
             throw new IOException("Closed");
 
-        write(s.getBytes(_channel.getResponse().getCharacterEncoding()));
+        // TODO would a Threadlocal pool be better?
+        String charset = _channel.getResponse().getCharacterEncoding();
+        if (_encoder==null || !_encoder.charset().name().equals(charset))
+            _encoder = Charset.forName(charset).newEncoder();
+        else
+            _encoder.reset();
+
+        CharBuffer in = CharBuffer.wrap(s);
+        CharBuffer crlf = eoln?CharBuffer.wrap("\r\n"):null;
+        ByteBuffer out = getHttpChannel().getByteBufferPool().acquire((int)(1+(s.length()+2)*_encoder.averageBytesPerChar()),false);
+        BufferUtil.flipToFill(out);
+
+        for(;;)
+        {
+            CoderResult result;
+            if (in.hasRemaining())
+            {
+                result = _encoder.encode(in, out, false);
+                if (result.isUnderflow())
+                    if (crlf==null)
+                        break;
```

Comment:
Not sure about this. If `crlf==null` we never call `encode(in, out, true)` - note the `true` parameter. Is that ok?

Reply:
You cannot pass `crlf==null` because there may be multiple encoding passes.

Reply:
I think I can.  On every call it is true that the input is the last input, even if they can only partially encode due to overflow.  Otherwise, how else can you tell it will be the last unless you know the encoding?